### PR TITLE
[RELENG-4763] verify VM spawn correctly

### PR DESCRIPTION
**Pull request by erwan-b**

### CODE REVIEW

File: runners_manager/vm_creation/openstack.py [hunk 1]
```diff
@@ -74,6 +75,12 @@ def create_vm(self, runner: Runner, runner_token: int or None,
             userdata=self.script_init_runner(runner, runner_token, github_organization,
                                              installer)
         )
+        while self.nova_client.servers.get(instance.id).status not in ['ACTIVE', 'ERROR']:
+            time.sleep(2)
+
+        if self.nova_client.servers.get(instance.id).status == 'ERROR':
+            logger.info('vm is got an error, try recreating')
```
Comment (tcarmet) [inline runners_manager/vm_creation/openstack.py:82]: Let's adjust this error message:
```suggestion
            logger.info('VM creation failed, creating a new one')
```

Also does `self.nova_client.server.get(instance.id)` gives us a description of the error if any? Could be nice to print it in the logs as well.

File: runners_manager/vm_creation/openstack.py [hunk 2]
```diff
@@ -74,6 +75,12 @@ def create_vm(self, runner: Runner, runner_token: int or None,
             userdata=self.script_init_runner(runner, runner_token, github_organization,
                                              installer)
         )
+        while self.nova_client.servers.get(instance.id).status not in ['ACTIVE', 'ERROR']:
+            time.sleep(2)
```
Comment (tcarmet) [inline runners_manager/vm_creation/openstack.py:79]: Open thought, do you think we should increment the sleep depending on the number of retries?

Comment (erwan-b) [reply runners_manager/vm_creation/openstack.py]: Well, if we want to reduce the number of api call, we could take the average spawn time and then start to check the status. (with an average fo 30 sec, it remove 15 calls already)

File: runners_manager/runner/RunnerManager.py [hunk 3]
```diff
@@ -65,6 +65,13 @@ def update(self, github_runners: list[dict]):
             while self.need_new_runner(vm_type):
                 self.create_runner(vm_type)
 
+            # Respawn runner if they are offline for more then 10min after spawn
+            for elem in self.filter_runners(vm_type, lambda r: r.status == 'offline'):
+                time_since_spawn = datetime.datetime.now() - elem.created_at
+                if elem.has_run is False \
+                        and time_since_spawn > datetime.timedelta(minutes=15):
+                    self.respawn_replace(elem)
```
Comment (tcarmet) [inline runners_manager/runner/RunnerManager.py:73]:
```suggestion
            for runner in self.filter_runners(vm_type, lambda r: r.status == 'offline'):
                time_since_spawn = datetime.datetime.now() - runner.created_at
                if runner.has_run is False \
                        and time_since_spawn > datetime.timedelta(minutes=15):
                    self.respawn_replace(runner)
```

File: runners_manager/runner/RunnerManager.py [hunk 4]
```diff
@@ -65,6 +65,13 @@ def update(self, github_runners: list[dict]):
             while self.need_new_runner(vm_type):
                 self.create_runner(vm_type)
 
+            # Respawn runner if they are offline for more then 10min after spawn
```
Comment (tcarmet) [inline runners_manager/runner/RunnerManager.py:68]: Your comment says 10 minutes, the code says 15.
```suggestion
            # Respawn runner if they are offline for more then 15min after spawn
```

File: runners_manager/runner/RunnerManager.py [hunk 5]
```diff
@@ -65,6 +65,13 @@ def update(self, github_runners: list[dict]):
...
+            for elem in self.filter_runners(vm_type, lambda r: r.status == 'offline'):
+                time_since_spawn = datetime.datetime.now() - elem.created_at
+                if elem.has_run is False \
+                        and time_since_spawn > datetime.timedelta(minutes=15):
```
Comment (tcarmet) [inline runners_manager/runner/RunnerManager.py:72]: This setting is pretty cool. we could add a TODO here to make it configurable one day. Maybe, maybe not :D

File: runners_manager/runner/tests/tests_runner.py [hunk 6]
```diff
@@ -95,9 +94,15 @@ def test_config_vm_type(self, name_generator: Mock,
         self.assertEqual(r.runner_management.__len__(), 1)
 
         self.assertEqual(r.runners.__len__(), 2)
-        self.assertEqual(r.runners['0'], Runner('0', '1', r.runner_management[0]))
-        self.assertEqual(r.runners['1'].__dict__,
-                         Runner('1', '2', r.runner_management[0]).__dict__)
+        # Runner('0', '1', r.runner_management[0])
```
Comment (tcarmet) [inline runners_manager/runner/tests/tests_runner.py:97]: Either remove or uncomment this line. Otherwise we will forget what to do with it :D

**Review by tcarmet — APPROVED** [review]
LGTM, but left some comments to be addressed

**Pull request merged**